### ISSUE TITLE
Is cyrb128 Public domain like cyrb53?

### ISSUE THREAD

**Issue by asl97** [bryc/code#13]
Is cyrb128 Public domain like cyrb53?

from your post on https://stackoverflow.com/a/47593316/1986995
```javascript
function cyrb128(str) {
    let h1 = 1779033703, h2 = 3144134277,
        h3 = 1013904242, h4 = 2773480762;
    for (let i = 0, k; i < str.length; i++) {
        k = str.charCodeAt(i);
        h1 = h2 ^ Math.imul(h1 ^ k, 597399067);
        h2 = h3 ^ Math.imul(h2 ^ k, 2869860233);
        h3 = h4 ^ Math.imul(h3 ^ k, 951274213);
        h4 = h1 ^ Math.imul(h4 ^ k, 2716044179);
    }
    h1 = Math.imul(h3 ^ (h1 >>> 18), 597399067);
    h2 = Math.imul(h4 ^ (h2 >>> 22), 2869860233);
    h3 = Math.imul(h1 ^ (h3 >>> 17), 951274213);
    h4 = Math.imul(h2 ^ (h4 >>> 19), 2716044179);
    return [(h1^h2^h3^h4)>>>0, (h2^h1)>>>0, (h3^h1)>>>0, (h4^h1)>>>0];
}
```

I am guessing this is how it is used
```javascript
function sfc32(a, b, c, d) {
    return function() {
      a >>>= 0; b >>>= 0; c >>>= 0; d >>>= 0; 
      var t = (a + b) | 0;
      a = b ^ b >>> 9;
      b = c + (c << 3) | 0;
      c = (c << 21 | c >>> 11);
      d = d + 1 | 0;
      t = t + d | 0;
      c = c + t | 0;
      return (t >>> 0) / 4294967296;
    }
}

get_random_gen = (seed)=>{
    return sfc32(...cyrb128(seed))
}

let random = get_random_gen('seed');
random()
```

**Comment by bryc** [comment]
yes, the whole PRNGs.md file linked there is public domain, i just didn't add cyrb128 to it yet.

**Comment by luciopaiva** [comment]
While reading this, I noticed that the version posted here by @asl97 differs from the one in the original SO thread. Instead of this:

```js
    return [(h1^h2^h3^h4)>>>0, (h2^h1)>>>0, (h3^h1)>>>0, (h4^h1)>>>0];
``` 

The original version does this:

```js
    h1 ^= (h2 ^ h3 ^ h4), h2 ^= h1, h3 ^= h1, h4 ^= h1;
    return [h1>>>0, h2>>>0, h3>>>0, h4>>>0]
``` 

Which is not the same!

I wonder if @asl97 changed the implementation or if it was copied from the SO thread before some recent edit by @bryc. Since they don't produce the same results, I'm curious to understand what happened.

**Comment by bryc** [comment]
<img width="1069" height="347" alt="Image" src="https://github.com/user-attachments/assets/3f52bade-ccc6-4016-b940-12ae14c91c29" />

Yeah that's an old revision. I had changed the final mixing procedure to reduce correlations (someone brought it up in comments as a potential flaw), as apparently it can be tricky to mix four 32-bit numbers to produce a 128-bit while ensuring the same entropy of an actual 128-bit number.

Since I saw it as just a seed generator, I never designed it to be good or maintain consistency. In fact I did another mixing improvement to it before this, and even before that, I entirely replaced the original example function I made called xmur3:

```js
function xmur3(str) {
    for(var i = 0, h = 1779033703 ^ str.length; i < str.length; i++) {
        h = Math.imul(h ^ str.charCodeAt(i), 3432918353);
        h = h << 13 | h >>> 19;
    } return function() {
        h = Math.imul(h ^ (h >>> 16), 2246822507);
        h = Math.imul(h ^ (h >>> 13), 3266489909);
        return (h ^= h >>> 16) >>> 0;
    }
}
```

This was ultimately just a crude PRNG with a lot of entropy built in, that gave you a new 32-bit seed each time you called it. Generally fine for producing seeds, I thought. Then I replaced it with a 128-bit version to potentially avoid any correlation issues when being used in PRNGs that take four 32-bit seeds like sfc32.

Although for serious purposes, I do recommend [MurmurHash3_x86_128](https://github.com/bryc/code/blob/master/jshash/hashes/murmurhash3_128.js) for a more well-tested and consistent 128-bit hash. Though it's designed to work with Array/TypedArray input and not strings.
